lib/cloud/gcp/vm.go

```diff
@@ -0,0 +1,545 @@
+	newKeys := strings.Join(existingKeys, "\n")
+	sshKeyItem.Value = &newKeys
+}
+
+// AddSSHKey adds an SSH key to a GCP VM's metadata.
+func (clt *instancesClient) AddSSHKey(ctx context.Context, req *SSHKeyRequest) error {
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	if req.PublicKey == nil {
+		return trace.BadParameter("public key not set")
+	}
+	addSSHKey(req.Instance.metadata, req.PublicKey, req.Expires)
+	op, err := clt.InstanceClient.SetMetadata(ctx, &computepb.SetMetadataInstanceRequest{
+		Instance:         req.Instance.Name,
+		MetadataResource: req.Instance.metadata,
+		Project:          req.Instance.ProjectID,
+		Zone:             req.Instance.Zone,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if err := op.Wait(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func removeSSHKey(meta *computepb.Metadata) {
+	for _, item := range meta.GetItems() {
+		if item.GetKey() != sshKeyName {
+			continue
+		}
+		existingKeys := strings.Split(item.GetValue(), "\n")
+		newKeys := make([]string, 0, len(existingKeys))
+		for _, key := range existingKeys {
+			if !strings.HasPrefix(key, sshUser) {
+				newKeys = append(newKeys, key)
+			}
+		}
+		item.Value = googleapi.String(strings.Join(newKeys, "\n"))
+		return
+	}
+}
+
+// RemoveSSHKey removes an SSH key from a GCP VM's metadata.
+func (clt *instancesClient) RemoveSSHKey(ctx context.Context, req *SSHKeyRequest) error {
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	removeSSHKey(req.Instance.metadata)
+	op, err := clt.InstanceClient.SetMetadata(ctx, &computepb.SetMetadataInstanceRequest{
+		Instance:         req.Instance.Name,
+		MetadataResource: req.Instance.metadata,
+		Project:          req.Instance.ProjectID,
+		Zone:             req.Instance.Zone,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if err := op.Wait(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// RunCommandRequest contains parameters for running a command on an instance.
+type RunCommandRequest struct {
+	// Client is the instance client to use.
+	Client InstancesClient
+	// InstanceRequest is the set of parameters identifying the instance.
+	InstanceRequest
+	// Script is the script to execute.
+	Script string
+	// SSHPort is the ssh server port to connect to. Defaults to 22.
+	SSHPort     string
+	dialContext func(ctx context.Context, network, addr string) (net.Conn, error)
+}
+
+func (req *RunCommandRequest) CheckAndSetDefaults() error {
+	if err := req.InstanceRequest.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	if req.Script == "" {
+		return trace.BadParameter("script must be set")
+	}
+	if req.SSHPort == "" {
+		req.SSHPort = "22"
+	}
+	if req.dialContext == nil {
+		dialer := net.Dialer{}
+		req.dialContext = dialer.DialContext
+	}
+	return nil
+}
+
+func generateKeyPair() (ssh.Signer, ssh.PublicKey, error) {
```

Comment:
Don't we already have a function somewhere that does this?

Reply:
We have some that generate raw key bytes, but none that I can find that also parse them.